go/common/subscription/new_heads_manager.go

```diff
@@ -93,18 +78,14 @@ func (nhs *NewHeadsService) Start() error {
 				if err != nil {
 					// on error, remove the notification
 					nhs.logger.Info("failed to notify newHead subscription", log.ErrKey, err, log.SubIDKey, id)
-					nhs.notifiersMutex.Lock()
 					delete(nhs.newHeadNotifiers, id)
-					nhs.notifiersMutex.Unlock()
 				}
 			}
 			return nil
 		},
 		func() {
-			if nhs.connectFunc == nil {
-				nhs.logger.Crit("the inbound new heads channel was closed.")
-			}
-			err := nhs.connectFunc(nhs.inputCh)
+			nhs.logger.Info("Disconnected from new head subscription. Reconnecting...")
+			inputCh, err = nhs.connectFunc()
```

Comment:
This assignment doesn't have the desired effect I don't think. It won't update the channel in the slice that was passed into ForwardFromChannels above.

To that you'd need to reference the slice, something like:
```
inputChans := []chan *common.BatchHeader{inputCh}

// and then down here you can write:			
inputCh, err = nhs.connectFunc()
inputChans[0] = inputCh
```

I wasn't sure how this worked, so did this to convince myself lol: https://go.dev/play/p/YD7x0HaUtoO

Reply:
Actually even that fix might be problematic I think (if I'm reading the magical incantations correctly).

The `inputCases` you build up with reflection at the start of `ForwardFromChannels()` will still be listening on the old channel.